example/hydra_plugins/launchers/my_submitit_launcher.py

```diff
@@ -0,0 +1,65 @@
+import logging
+from time import sleep
+from typing import Any
+
+import wandb
+from hydra_plugins.hydra_submitit_launcher.submitit_launcher import BaseSubmititLauncher
+
+logger = logging.getLogger(__name__)
+
+# This has to be located under a hydra_plugins folder or else hydra won't recognize this as a valid plugin.
+# Furthermore, placing an __init__.py under hydra_plugins will cause circular import issues with the above
+# 'from hydra_plugins' import (which is why hydra says not to do this
+# https://hydra.cc/docs/next/advanced/plugins/develop/#internaldocs-banner). However, although Hydra's instantiate
+# function can resolve classes without requiring __init__.py, submitit requires this when pickling. Otherwise
+# submitit won't be able to find the my_submitit_launcher module. A way around this is to make a nested folder,
+# 'launchers', with an __init__.py and place the module there. Then from train.py we register
+# hydra_plugins.launchers by value.
+class MyBaseSubmititLauncher(BaseSubmititLauncher):
+    def __init__(self, **params: Any) -> None:
+        super().__init__(**params)
+
+    def checkpoint(self, *args: Any, **kwargs: Any) -> Any:
```

Comment:
Not much aware of submitit functionalities. Does this checkpoint only get called when you want a cluster node preempties? That's because, I can see in some of the lines, that you are enforcing the run to restart when this function gets called. Might be problematic if this gets called in different circumstances (i.e., non-preempting cases)

Reply:
More specifically, it gets called when `SIGTERM` and/or `USR1` and/or `SIGCONT` are sent to the parent process and _only_ in a SLURM environment (it doesn't get used when using submitit_local since that's a local environment). It does not get accidentally called by wandb, if that's what you're worried about :) 

This can either happen from job timeouts when the process hasn't finished yet the allotted time has finished: in this case, SLURM signals `USR1` 120 seconds (#seconds can be changed) before the job times out (read about `USR1` coming from `sbatch` [here](https://slurm.schedmd.com/sbatch.html#OPT_signal)), then `SIGCONT`, then `SIGTERM`, then finally `SIGKILL` when the job has timed out. `checkpoint` gets called on `USR1` in this case and does the necessary cleanup and job resubmission before exiting the process itself (before `SIGKILL` arrives from SLURM). In the pre-emption case, there is no `USR1` and instead `SIGCONT` (and immediately afterwards `SIGTERM`) arrive, which gets caught by submitit and executes `checkpoint`, which gives it time to clean up, resubmit, and exit before `SIGKILL` arrives. You can read more about pre-emption under "Grace Time" [here](https://slurm.schedmd.com/preempt.html).

Just to make it clear, this does not explicitly restart the run since that's impossible on wandb's end at the moment. Wandb sweeps don't allow you to restart specific runs, which would've been really nice. Once this checkpoint function has marked the run as pre-emptible and finishes with a non-zero exit code, there's a chance that any of the other sweep agents currently running could pick up the requeued run (i.e., the same run parameters) before the checkpoint function has successfully resubmitted the task function (in this case, `__call__` in `DummyTraining`) and that task function has begun a sweep agent to pull runs from the sweep controller. Basically, `mark_preempting` just puts the run params back up on a global queue that ALL agents can pull from. What would be really nice would be if:
1. You can restrict wandb agents from pulling requeued runs from the wandb sweep controller. That way, when you specify `mark_preempting` during pre-emption, you can configure the next agent that will be a part of the resubmitted task_function to be able to pull the requeued run from the controller.
2. You can specify the run ID of the run to be initialized and if that run ID was an ID used in a pre-empted run, it'll pull the sweep params that were associated with that ID. That way, when you resubmit your task function during pre-emption, once it spins up its agent, it'll know exactly which run to continue from.